codegen-server/src/main/kotlin/software/amazon/smithy/rust/codegen/server/smithy/protocols/ServerProtocolLoader.kt

```diff
@@ -0,0 +1,45 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0.
+ */
+
+package software.amazon.smithy.rust.codegen.server.smithy.protocols
+
+import software.amazon.smithy.aws.traits.protocols.RestJson1Trait
+import software.amazon.smithy.codegen.core.CodegenException
+import software.amazon.smithy.model.Model
+import software.amazon.smithy.model.knowledge.ServiceIndex
+import software.amazon.smithy.model.shapes.ServiceShape
+import software.amazon.smithy.model.shapes.ShapeId
+import software.amazon.smithy.model.traits.Trait
+import software.amazon.smithy.rust.codegen.server.smithy.generators.protocol.ServerProtocolGenerator
+import software.amazon.smithy.rust.codegen.smithy.protocols.ProtocolGeneratorFactory
+import software.amazon.smithy.rust.codegen.smithy.protocols.ProtocolMap
+
+/*
+ * Protocol dispatcher, responsible for protocol selection.
+ */
+class ServerProtocolLoader(private val supportedProtocols: ProtocolMap) {
+    fun protocolFor(
+        model: Model,
+        serviceShape: ServiceShape
+    ): Pair<ShapeId, ProtocolGeneratorFactory<ServerProtocolGenerator>> {
+        val protocols: MutableMap<ShapeId, Trait> = ServiceIndex.of(model).getProtocols(serviceShape)
+        val matchingProtocols =
+            protocols.keys.mapNotNull { protocolId -> supportedProtocols[protocolId]?.let { protocolId to it } }
+        if (matchingProtocols.isEmpty()) {
+            throw CodegenException("No matching protocol — service offers: ${protocols.keys}. We offer: ${supportedProtocols.keys}")
+        }
+        val pair = matchingProtocols.first()
+        // TODO: is there a better way than an unsafe cast here?
```

Comment:
This is happening because `ProtocolMap` is defined as

```kotlin
typealias ProtocolMap = Map<ShapeId, ProtocolGeneratorFactory<ProtocolGenerator>>
```

and this function returns a pair whose second component is of type `ProtocolGeneratorFactory<ServerProtocolGenerator>`. So the cast is an unsafe cast casting `ProtocolGeneratorFactory<ProtocolGenerator>` to `ProtocolGeneratorFactory<ServerProtocolGenerator>`.

---

We can define our own

```
typealias ServerProtocolMap = Map<ShapeId, ProtocolGeneratorFactory<ServerProtocolGenerator>>
```

in this file, or write the type in full in the signature of this function. The cast will then become unnecessary.

Reply:
Thanks for teaching me some Kotlin 😄 !! I'll do this.